File: frontend/src/api/registry.ts
import { apiClient } from '@/api/client';

export interface RegistryLookupResult {
  egid: number;
  source: string;
  address: string | null;
  postal_code: string | null;
  city: string | null;
  canton: string | null;
  construction_year: number | null;
  building_category: string | null;
  building_class: string | null;
  floors: number | null;
  area: number | null;
  heating_type: string | null;
  energy_source: string | null;
  renovation_year: number | null;
  coordinates: { lat: number; lng: number } | null;
  raw_attributes: Record<string, unknown>;
}

export interface AddressSearchResult {
  source: string;
  address: string | null;
  postal_code: string | null;
  city: string | null;
  canton: string | null;
  lat: number | null;
  lng: number | null;
  egid: number | null;
  feature_id: string | null;
}

export interface HazardLevel {
  level: string;
  description: string | null;
  source: string | null;
}

export interface NaturalHazardsResult {
  flood_risk: HazardLevel | null;
  landslide_risk: HazardLevel | null;
  avalanche_risk: HazardLevel | null;
  earthquake_risk: HazardLevel | null;
}

export interface EnrichmentResult {
  building_id: string;
  updated_fields: Record<string, unknown>;
  source: string;
  egid_found: boolean;
  hazards_fetched: boolean;
}

export const registryApi = {
  lookupByEgid: async (egid: number): Promise<RegistryLookupResult> => {
    const response = await apiClient.get<RegistryLookupResult>(`/registry/lookup/egid/${egid}`);
    return response.data;
  },

  searchByAddress: async (query: string, postalCode?: string): Promise<AddressSearchResult[]> => {
    const params: Record<string, string> = { q: query };
    if (postalCode) params.postal_code = postalCode;
    const response = await apiClient.get<AddressSearchResult[]>('/registry/lookup/address', {
      params,
    });
    return response.data;
  },

  getNaturalHazards: async (lat: number, lng: number): Promise<NaturalHazardsResult> => {
    const response = await apiClient.get<NaturalHazardsResult>('/registry/hazards', {
      params: { lat, lng },
    });
    return response.data;
  },

  enrichBuilding: async (buildingId: string): Promise<EnrichmentResult> => {
    const response = await apiClient.post<EnrichmentResult>(`/buildings/${buildingId}/enrich`);
    return response.data;
  },
};

File: backend/app/schemas/registry.py
"""Schemas for Swiss public registry lookups and building enrichment."""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel


class RegistryLookupResult(BaseModel):
    """Result of a RegBL EGID lookup."""

    egid: int
    source: str = "regbl"
    address: str | None = None
    postal_code: str | None = None
    city: str | None = None
    canton: str | None = None
    construction_year: int | None = None
    building_category: str | None = None
    building_class: str | None = None
    floors: int | None = None
    area: float | None = None
    heating_type: str | None = None
    energy_source: str | None = None
    renovation_year: int | None = None
    coordinates: dict[str, float] | None = None
    raw_attributes: dict[str, Any] = {}


class AddressSearchResult(BaseModel):
    """Single result from a Swisstopo address search."""

    source: str = "swisstopo"
    address: str | None = None
    postal_code: str | None = None
    city: str | None = None
    canton: str | None = None
    lat: float | None = None
    lng: float | None = None
    egid: int | None = None
    feature_id: str | None = None


class HazardLevel(BaseModel):
    """A single hazard risk level."""

    level: str = "unknown"
    description: str | None = None
    source: str | None = None


class NaturalHazardsResult(BaseModel):
    """Natural hazards at a geographic point."""

    flood_risk: HazardLevel | None = None
    landslide_risk: HazardLevel | None = None
    avalanche_risk: HazardLevel | None = None
    earthquake_risk: HazardLevel | None = None


class EnrichmentResult(BaseModel):
    """Result of auto-enriching a building from public registries."""

    building_id: str
    updated_fields: dict[str, Any] = {}
    source: str = "regbl+geo.admin"
    egid_found: bool = False
    hazards_fetched: bool = False
